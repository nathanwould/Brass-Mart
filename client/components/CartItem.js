import { Card } from "antd";
import formatMoney from "../lib/formatMoney";
import RemoveFromCart from "./RemoveFromCart";

export default function CartItem({ cartItem, product }) {
  const { Meta } = Card;
  console.log(product)
  if (!product) return null;
  return (
    <Card
    cover={
      <img
        alt={product.name}
          src={product.photos[0].image.publicUrlTransformed}
      />
    }
    >
      <Meta
        title={product.name}
        description={formatMoney(product.price)}
      />
      <RemoveFromCart id={cartItem.id}/>
    </Card>
  )
}