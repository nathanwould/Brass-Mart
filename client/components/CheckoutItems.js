import { Card, Space } from "antd";
import formatMoney from "../lib/formatMoney";
import calcTotalPrice from "../lib/calcTotalPrice";
import RemoveFromCart from "./RemoveFromCart";

export default function CheckoutItems({ cart }) {
  const { Meta } = Card;
  if (!cart) {
    return (
      <p>Loading...</p>
    )
  }
  return (
    <>
      <Space
        direction="vertical"
        // align="center"
        style={{
          padding: "2vh",
          background: "white",
          borderRadius: "2%",
          marginBottom: "4vh"
        }}
      >
        {cart?.map(cartItem => (
          <Space
            key={cartItem.id}
            align="center"
            style={{
              margin: "1vh"
            }}
          >
          <img
            alt={cartItem.product.name}
              src={cartItem.product.photos[0].image.publicUrlTransformed}
              style={{
                width: "10vh"
              }}
            />
          <p>{cartItem.product.name}</p>
          <p>{formatMoney(cartItem.product.price)}</p>
          <RemoveFromCart id={cartItem.id} />
        </Space>
        ))}
        <p style={{
          margin: "1vh",
          fontWeight: "bold"
        }}
        >
          Total: {formatMoney(calcTotalPrice(cart))}
        </p>
        </Space>
      </>
  )
}