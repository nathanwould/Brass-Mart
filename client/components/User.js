import { gql, useQuery } from '@apollo/client';

const CURRENT_USER_QUERY = gql`
  query {
    authenticatedItem {
      ... on User {
        id
        email
        name
        cart {
          id
          quantity
          product {
            id
            price
            name
            photos {
              image {
                publicUrlTransformed
              }
            }
          }
        }
        orders {
          id
          total
          charge
          items {
            id
            name
            description
            price
            quantity
            photos {
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`;

export function useUser() {
  const { data } = useQuery(CURRENT_USER_QUERY);
  console.log(data?.authenticatedItem);
  return data?.authenticatedItem;
}

export { CURRENT_USER_QUERY };
