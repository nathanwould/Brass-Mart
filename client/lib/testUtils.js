import casual from 'casual';

casual.seed(777);

const fakeProduct = () => ({
  id: 'abc123',
  price: 50000,
  name: 'The Sickest Trombone',
  make: 'Huffy',
  model: 'Big Boy',
  productType: 'instrument',
  category: 'trombone',
  description: 'A wicked sick product!',
  photos: [{
    id: 123,
    altText: 'a wicked sick product',
    image: {
      publicUrlTransformed: 'product.jpg',
    },
  }],
  status: 'IN-STOCK',
});

const fakeUser = (overrides) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  orders: [],
  cart: [],
  cartCount: 2,
  ...overrides,
});

const fakeCartItem = (overrides) => ({
  __typename: 'CartItem',
  id: '123456',
  product: fakeProduct(),
  quantity: 1,
  user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  fakeProduct,
  fakeUser,
  fakeCartItem,
  LocalStorageMock
};
