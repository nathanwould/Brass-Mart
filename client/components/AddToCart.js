import gql from 'graphql-tag';
import { useMutation } from "@apollo/client";
import { CURRENT_USER_QUERY, useUser } from './User';
import { Button } from 'antd';

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(productId: $id) {
      id
    }
  }
`;

export default function AddToCart({ id }) {
  const user = useUser();
  const added = !!user?.cart.find(cartItem => cartItem.product.id === id)
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id },
    refetchQueries: [{ query: CURRENT_USER_QUERY }]
  });
  return (
    <>
      {added ?
        <Button
          disabled
          type="button">
          Added To Cart
        </Button>
        : <Button
          type="button"
          onClick={addToCart}
          id={id}>
          {`Add${loading ? 'ing' : ''} To Cart ðŸ›’`}
        </Button>
      }
    </>
  );
}
