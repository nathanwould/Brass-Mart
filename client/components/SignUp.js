import { useMutation } from "@apollo/client";
import { gql } from "graphql-tag";
import { Button, Form, Input, Typography } from 'antd';
import { CURRENT_USER_QUERY, useUser } from "./User";
import React from "react";
import { UserOutlined, MailOutlined, LockOutlined } from '@ant-design/icons';

export const SIGN_UP_MUTATION = gql`
  mutation SIGN_UP_MUTATION(
    $name: String!,
    $email: String!,
    $password: String!
  ) {
    createUser(data: {name: $name, email: $email, password: $password}) {
      id
      email
      name
    }
  }
`;

export default function SignUp() {
  const formRef = React.createRef();
  const Title = Typography;
  const [signup, { data, loading }] = useMutation(SIGN_UP_MUTATION);
  const onFinishFailed = (errorInfo) => {
    console.log('Failed:', errorInfo);
  };
  const user = useUser();
  return (
    <div className="form-div"
      style={{
        margin: "6vh auto",
        display: "flex",
        flexDirection: "column",
        justifyContent: "center",
        alignItems: "center",
      }}
    >
      {!user ? <h2>Sign Up</h2>
        : <h2>Welcome, {user.name}</h2>
      }
      <Form
        ref={formRef}
        size="middle"
        name='control-ref'
        // labelCol={{ span: 8 }}
        // wrapperCol={{ span: 16 }}
        initialValues={{ remember: true }}
        onFinish={(values) => (
          signup({
            variables: values,
            refetchQueries: [{ query: CURRENT_USER_QUERY }],
          })
        )}
        onFinishFailed={onFinishFailed}
        style={{
          width: '50%',
          textAlign: 'center'
        }}
      >
        <Form.Item
          name="name"
          rules={[
            {
              required: true,
              message: 'Please enter your name'
            },
          ]}
        >
          <Input
            prefix={<UserOutlined className="site-form-item-icon" />}
            type="name"
            placeholder="Your Name"
            autoComplete="name"
            disabled={user}
          />
        </Form.Item>
        <Form.Item
          name="email"
          rules={[
            {
              required: true,
              message: 'Please enter your Email'
            },
          ]}
        >
          <Input
            prefix={<MailOutlined className="site-form-item-icon" />}
            type="email"
            placeholder="Your Email Address"
            autoComplete="email"
            disabled={user}
          />
        </Form.Item>
        <Form.Item
          name="password"
          rules={[
            {
              required: true,
              message: 'Please enter your password'
            }
          ]}
        >
          <Input.Password
            prefix={<LockOutlined className="site-form-item-icon" />}
            type="password"
            placeholder="Password"
            autoComplete="password"
            disabled={user}
          />
        </Form.Item>
        <Form.Item>
          <Button
            type="primary"
            htmlType="submit"
            className="login-form-button"
            data-testid="sign-up-button"
            disabled={user}
            style={{
              width: "100%"
            }}
          >
            {user ? <span>Signed In!</span>
              : <span>Sign Up</span>}
          </Button>
          Or <a href="signin">Sign In</a>
        </Form.Item>
      </Form>
    </div>
  );
}
