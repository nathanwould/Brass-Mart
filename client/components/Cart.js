import { Button, Drawer, Layout } from "antd";
import { useCart } from "../lib/cartState";
import formatMoney from "../lib/formatMoney";
import calcTotalPrice from "../lib/calcTotalPrice";
import CartItem from "./CartItem";
import { useUser } from "./User";

export default function Cart() {
  const user = useUser();
  const { cartOpen, closeCart } = useCart();
  const { Content, Footer } = Layout;

  return (
    <Drawer
      title={`${user?.name}'s Cart`}
      placement="right"
      onClose={closeCart}
      visible={cartOpen}
    >
      <Content style={{
        height: "75vh",
        overflow: "scroll"
      }}>
        {user?.cart?.map(cartItem => (
          <CartItem
            key={cartItem.id}
            cartItem={cartItem}
            product={cartItem.product} />
        ))
        }
      </Content>
      <Footer style={{
        alignItems: "center",
        justifyContent: "center",
        padding: "2vh"
      }}>
          {user && <p>Total: {formatMoney(calcTotalPrice(user?.cart))}</p>}
          <Button type="primary" href="/checkout">Checkout</Button>
        </Footer>
    </Drawer>
  )
}