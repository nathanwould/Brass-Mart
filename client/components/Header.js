import Link from 'next/link';
import Cart from './Cart';
import Nav from './Nav';

export default function Header({ Layout }) {
  const { Header } = Layout;
  return (
    <Header style={{ padding: 0 }}>
      <Nav Layout={Layout }/>
      <Cart />
    </Header>
  );
}
