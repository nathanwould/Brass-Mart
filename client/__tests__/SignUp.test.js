import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import wait from 'waait';
import SignUp, { SIGN_UP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';
import gql from 'graphql-tag';

const me = fakeUser();
const password = 'BosDaddy';
const mocks = [
  {
    request: {
      query: SIGN_UP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password,
      },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'abc123',
          name: me.name,
          email: me.email,
        },
      },
    },
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: me } },
  },
];

// current user mock
// const userMocks = [
//   {
//     request: { query: CURRENT_USER_QUERY },
//     result: { data: { authenticatedItem: me } },
//   },
// ];

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

describe('<SignUp />', () => {
  it('renders and matches snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it('calls the mutation properly', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>
    );
    // checking typing into boxes works
    await userEvent.type(screen.getByPlaceholderText(/name/i), me.name);
    await userEvent.type(screen.getByPlaceholderText(/email/i), me.email);
    await userEvent.type(screen.getByPlaceholderText(/password/i), password);
    // Click the sign up button
    await userEvent.click(screen.getByTestId('sign-up-button'));
    wait(400);
    await screen.findByText(
      `Welcome, ${me.name}`
    );
  });
});