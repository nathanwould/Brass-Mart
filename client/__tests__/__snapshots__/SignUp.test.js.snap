// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches snapshot 1`] = `
<div>
  <div
    class="form-div"
    style="margin: 6vh auto; display: flex; flex-direction: column; justify-content: center; align-items: center;"
  >
    <h2>
      Sign Up
    </h2>
    <form
      class="ant-form ant-form-horizontal ant-form-middle"
      id="control-ref"
      style="width: 50%; text-align: center;"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <span
                class="ant-input-affix-wrapper"
              >
                <span
                  class="ant-input-prefix"
                >
                  <span
                    aria-label="user"
                    class="anticon anticon-user site-form-item-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="user"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                      />
                    </svg>
                  </span>
                </span>
                <input
                  autocomplete="name"
                  class="ant-input"
                  id="control-ref_name"
                  placeholder="Your Name"
                  type="name"
                  value=""
                />
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <span
                class="ant-input-affix-wrapper"
              >
                <span
                  class="ant-input-prefix"
                >
                  <span
                    aria-label="mail"
                    class="anticon anticon-mail site-form-item-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="mail"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0068.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                      />
                    </svg>
                  </span>
                </span>
                <input
                  autocomplete="email"
                  class="ant-input"
                  id="control-ref_email"
                  placeholder="Your Email Address"
                  type="email"
                  value=""
                />
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <span
                class="ant-input-affix-wrapper ant-input-password"
              >
                <span
                  class="ant-input-prefix"
                >
                  <span
                    aria-label="lock"
                    class="anticon anticon-lock site-form-item-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="lock"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 464h-68V240c0-70.7-57.3-128-128-128H388c-70.7 0-128 57.3-128 128v224h-68c-17.7 0-32 14.3-32 32v384c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V496c0-17.7-14.3-32-32-32zM332 240c0-30.9 25.1-56 56-56h248c30.9 0 56 25.1 56 56v224H332V240zm460 600H232V536h560v304zM484 701v53c0 4.4 3.6 8 8 8h40c4.4 0 8-3.6 8-8v-53a48.01 48.01 0 10-56 0z"
                      />
                    </svg>
                  </span>
                </span>
                <input
                  action="click"
                  autocomplete="password"
                  class="ant-input"
                  id="control-ref_password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <span
                  class="ant-input-suffix"
                >
                  <span
                    aria-label="eye-invisible"
                    class="anticon anticon-eye-invisible ant-input-password-icon"
                    role="img"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="eye-invisible"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 000-51.5zm-63.57-320.64L836 122.88a8 8 0 00-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 000 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 000 11.31L155.17 889a8 8 0 0011.31 0l712.15-712.12a8 8 0 000-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 00-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 01146.2-106.69L401.31 546.2A112 112 0 01396 512z"
                      />
                      <path
                        d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 00227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 01-112 112z"
                      />
                    </svg>
                  </span>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <button
                class="ant-btn ant-btn-primary login-form-button"
                data-testid="sign-up-button"
                style="width: 100%;"
                type="submit"
              >
                <span>
                  Sign Up
                </span>
              </button>
              Or 
              <a
                href="signin"
              >
                Sign In
              </a>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
